sageresearch-app-sdk/src/main/java/org/sagebionetworks/research/sageresearch_app_sdk/archive/TaskResultArchiveFactory.java

```diff
@@ -26,12 +34,36 @@ public boolean isSupported(@NonNull final Result result) {
     @Override
     public ImmutableSet<ArchiveFile> toArchiveFiles(@NonNull final Result result) {
         ImmutableSet.Builder<ArchiveFile> builder = new ImmutableSet.Builder<>();
-        for (Result stepResult : ((TaskResult) result).getStepHistory()) {
+
+        if (!(result instanceof TaskResult)) {
+            LOGGER.error("Result " + result.getIdentifier() + " must be a TaskResult.");
+            return builder.build();
+        }
+        TaskResult taskResult = (TaskResult)result;
+
+        // Create archives for the step history
+        List<Result> stepHistory = taskResult.getStepHistory();
+        for (Result stepResult: stepHistory) {
             builder.add(abstractResultArchiveFactory.toArchiveFiles(stepResult).toArray(new ArchiveFile[]{}));
         }
-        for (Result asynResult : ((TaskResult) result).getAsyncResults()) {
+
+        // Create the archives for the async results
+        List<Result> asyncResults = taskResult.getAsyncResults();
+        for (Result asynResult : asyncResults) {
             builder.add(abstractResultArchiveFactory.toArchiveFiles(asynResult).toArray(new ArchiveFile[]{}));
         }
+
+        // Create the archives for the result list archive factories
+        if (abstractResultArchiveFactory instanceof AbstractResultListArchiveFactory) {

```

Comment:
Having toArchiveFiles(@NonNull final ImmutableList<Result> resultList) implemented by SageResearchResultArchiveFactory muddles the AbstractResultArchiveFactory pattern. I'm thinking having TaskResultArchiveFactory use AnswerMapResultListFactory (or ResultListFactory) would be clearer.

Reply:
I agree with all your comments.  I switched over to the SageResearchResultArchiveFactory to take in an injected TaskResultArchiveFactory that the app can choose to provide the base one or a sub-class I made called TaskResultAnswerMapArchiveFactory that adds the answer map json archive along with the same functionality it had before.  Check out the code now and lmk if you think this is better.

Also, I think this will create a problem when we move to multiple nested TaskResults. I think each task result will create their own "answers" json archive and only the last one in the list will be upload, we may need some sort of archive merging functionality when we start doing that.